File: Client Retention.py
#!/usr/bin/env python
# coding: utf-8

# In[4]:


#Load Libraries
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency
import seaborn as sns
import knn_impute as ki


# In[3]:


#Set working directory
os.chdir("E:\\Notes and ppts\\Edwisor\\Projects\\Project 3")


# In[167]:


#Loading Data
train = pd.read_csv("Train_data.csv")
test = pd.read_csv("Test_data.csv")


# In[168]:


#Appending train and test data
data = train.append(test)


# In[169]:


data.head()


# In[170]:


#Data Preprocessing
# Encoding the Independent categorical Variable
from sklearn.preprocessing import LabelEncoder
labelencoder_X = LabelEncoder()
data.iloc[:, 0] = labelencoder_X.fit_transform(data.iloc[:, 0])
data.iloc[:, 2] = labelencoder_X.fit_transform(data.iloc[:, 2])
data.iloc[:, 3] = labelencoder_X.fit_transform(data.iloc[:, 3])
data.iloc[:, 4] = labelencoder_X.fit_transform(data.iloc[:, 4])
data.iloc[:, 5] = labelencoder_X.fit_transform(data.iloc[:, 5])
data.iloc[:, 20] = labelencoder_X.fit_transform(data.iloc[:, 20])

#Converting the respective data type to categorical
data.state = data.state.astype('category')
data['international plan'] = data['international plan'].astype('category')
data['voice mail plan'] = data['voice mail plan'].astype('category')
data.Churn = data.Churn.astype('category')
data['area code'] = data['area code'].astype('category')
data['phone number'] = data['phone number'].astype('category')


# In[171]:


data.dtypes


# In[301]:


#Checking for missing value
data.isnull().sum()


# In[145]:


#Visualizing box plot
get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(data['total day charge'])


# In[172]:


#Extracting Numeric and Categorical data
numeric_data = data.select_dtypes(exclude = ['category'])
factor_data = data.select_dtypes(include=['category'])


# In[173]:


cnames = numeric_data.columns


# In[174]:


#Checking Outliers and replacing them with NAs
for i in cnames:
    #calculate quartiles
    q75, q25 = np.percentile(numeric_data[i], [75 ,25])
    #Calculate Inter-quartile range
    iqr = q75 - q25
    #Calculate limits
    mini = q25 - (iqr*1.5)
    maxi = q75 + (iqr*1.5)
    #Replace with NA
    numeric_data.loc[numeric_data[i] < mini,:i] = np.nan
    numeric_data.loc[numeric_data[i] > maxi,:i] = np.nan


# In[176]:


#imputing NAs using KNN method
for i in cnames:
    ki.knn_impute(numeric_data[i],numeric_data,7)


# In[177]:


#Confirming the imputation
data.isna().sum()


# In[15]:


#Set the width and hieght of the plot
f, ax = plt.subplots(figsize=(7, 5))

#Generate correlation matrix
corel = numeric_data.corr()

#Ploting the correlation analysis
sns.heatmap(corel, mask=np.zeros_like(corel, dtype=np.bool), cmap=sns.diverging_palette(220, 10, as_cmap=True),
            square=True, ax=ax)


# In[196]:


cate_var = data.columns


# In[197]:


#Check Chi Square Analysis
for i in cate_var:
    print(i)
    chi2, p, dof, ex = chi2_contingency(pd.crosstab(data['Churn'], data[i]))
    print(p, chi2)


# In[58]:


#combining numeric and factor data
data = pd.concat([numeric_data, factor_data],axis = 1)


# In[180]:


#Dimensionality reduction
data = data.drop(['area code','phone number','total intl minutes','total day minutes','total eve minutes','total night minutes'],axis = 1)


# In[181]:


#Splitting the train and test data in original ratio
train = data.iloc[0:3333,:]
test = data.iloc[3333:5000,:]


# In[182]:


#Model Development
from sklearn import tree
from sklearn.metrics import accuracy_score


# In[183]:


data.shape


# In[184]:


#Splitting the dependent and independent from test and train set
x_train = train.values[:,0:14]
y_train = train.values[:,14]
x_test = test.values[:,0:14]
y_test = test.values[:,14]


# In[185]:


#converting the data type of the dependent variable train and test data
y_train = pd.Series(y_train).astype('category').values
y_test = pd.Series(y_test).astype('category').values


# In[186]:


#Decision Tree Model
DT_model = tree.DecisionTreeClassifier(criterion='entropy').fit(x_train, y_train)

#Predicting the test dataset
DT_Predictions = DT_model.predict(x_test)


# In[271]:


#Confusion matrix
from sklearn.metrics import confusion_matrix 
CM1 = confusion_matrix(y_test, DT_Predictions)
CM = pd.crosstab(y_test, DT_Predictions)

#Saving True-Positive, True-Negative, False-Positive, False-Negative
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
print(((TP+TN)*100)/(TP+TN+FP+FN))

#False Negative rate 
print((FN*100)/(FN+TP))

#Results
#Accuracy: 92.14
#FNR: 32.58


# In[272]:


CM


# In[ ]:


#Building Random Forest Model
from sklearn.ensemble import RandomForestClassifier

RF_model = RandomForestClassifier(n_estimators = 250).fit(x_train, y_train)


# In[189]:


#Predicting the dependent variable
RF_Predictions = RF_model.predict(x_test)


# In[269]:


#Confusion Matrix
# from sklearn.metrics import confusion_matrix 
CM1 = confusion_matrix(y_test, RF_Predictions)
CM = pd.crosstab(y_test, RF_Predictions)

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
print(((TP+TN)*100)/(TP+TN+FP+FN))

#False Negative rate 
print((FN*100)/(FN+TP))

#Accuracy: 95.68
#FNR: 29.46


# In[191]:


#Building Naive Bayes Model
from sklearn.naive_bayes import GaussianNB

#Naive Bayes implementation
NB_model = GaussianNB().fit(x_train, y_train)


# In[192]:


#predict test cases
NB_Predictions = NB_model.predict(x_test)


# In[273]:


#Confusion matrix
CM = pd.crosstab(y_test, NB_Predictions)
CM1 = confusion_matrix(y_test, NB_Predictions)
#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
print(((TP+TN)*100)/(TP+TN+FP+FN))

#False Negative rate 
print((FN*100)/(FN+TP))

#Accuracy: 85.84
#FNR: 60.26


# In[ ]:

File: Client Retention.R
#Setting Working Directory
setwd("E:/Notes and ppts/Edwisor/Projects/Project 3")

#Loading train and test data
train = read.csv("Train_data.csv")
test = read.csv("Test_data.csv")


#converting the data type
train$area.code = as.factor(train$area.code)
test$area.code = as.factor(test$area.code)

#Combining Train and test data
data = rbind(train, test)

#Exploring the data
str(data)

#Missing Value Analysis
sum(is.na(data))


#splitting data into factor and numeric type
factor.index = sapply(data, is.factor)
numeric.index = sapply(data, is.numeric)
factor.data = data[,factor.index]
numeric.data = data[,numeric.index]


#Outlier Analysis
cnames = colnames(numeric.data)
for (i in 1:length(cnames))
{
  assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "Churn"), data = subset(data))+
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="red", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i],x="Churn")+
           ggtitle(paste("Box plot of Churn for",cnames[i])))
}

## Plotting plots together
gridExtra::grid.arrange(gn1,gn2,gn3,ncol=3)
gridExtra::grid.arrange(gn4,gn5,gn6,ncol=3)
gridExtra::grid.arrange(gn7,gn8,gn9,ncol=3)
gridExtra::grid.arrange(gn10,gn11,gn12,ncol=3)
gridExtra::grid.arrange(gn13,gn14,gn15,ncol=3)

#replacing all the outliers with NA and imputing them using knn method


columns = colnames(data)
for(i in columns){
    val = data[,i][data[,i] %in% boxplot.stats(data[,i])$out]
    print(length(val))
    data[,i][data[,i] %in% val] = NA
  }
sum((is.na(data)))

#number of outliers detected and removed = 1080
library(DMwR)
data = knnImputation(data, k = 5)

#Detecting correalation and dependence between variables 

#plotting correlogram
corrgram(numeric.data, order = FALSE, upper.panel = panel.pie, text.panel = panel.txt, main = "Correlation Plot")
plot(data$total.day.minutes, data$total.day.charge)

corre_ana = cor(numeric.data, numeric.data)
write.csv(corre_ana, "Correlation Analysis.csv")

plot(data2$total.eve.minutes,data2$total.eve.charge, xlab = "Total Evening Minutes",ylab = "Total Evening Charge")

#Chi Square Test of Independence

for (i in 1:21)
{
  print(names(data)[i])
  print(chisq.test(table(factor.data$Churn,data[,i])))
}


#Dimensionality Reduction
data = subset(data, select = -c(area.code,phone.number,total.day.minutes,total.eve.minutes,total.night.minutes))

#recreating original train and test data partition

train = data[1:3333,]
test = data[3334:5000,]

#Creating Model

#Decision Tree Model
library(C50)
library(rpart)
classifier = C5.0(Churn~., train, trials = 100, rules = TRUE)
summary(classifier)
write(capture.output(summary(classifier)), "C50 Output.txt")
regressor = rpart(Churn~., data = train)
C50_predict = predict(classifier, test[,-16], type = "class")
table(C50_predict)
table(test$Churn)

#Plotting Decision Tree and result
regressor = rpart(Churn~., data = train)
rpart.plot::rpart.plot(regressor, box.palette = "RdBu", shadow.col = "gray", nn=TRUE)
plot(C50_predict, test[,16], xlab = "Predicted Churn Score", ylab = "Actual Churn Score")


##Evaluate the performance of classification model
ConfMatrix_C50 = table(test$Churn, C50_predict)
library(caret)
confusionMatrix(ConfMatrix_C50)

#Random Forest Model
library(randomForest)
RF_model = randomForest(Churn~.,data = train, ntree = 100)
pred = predict(RF_model, test[,-16], type = "class")
table(pred)

#Extract rules fromn random forest
#transform rf object to an inTrees' format
library(inTrees)
treeList = RF2List(RF_model) 
 
#Extract rules
exec = extractRules(treeList, train[,-16])
 
#Make rules more readable:
readableRules = presentRules(exec, colnames(train))

#Get rule metrics
ruleMetric = getRuleMetric(exec, train[,-16], train$responded)  # get rule metrics

#Writing rules to a text file
write(capture.output(readableRules), "Random Forest Rules.txt")

#Get rule metrics
ruleMetric = getRuleMetric(exec, train[,-16], train$Churn)
write(capture.output(ruleMetric), "Random Forest Rule metric.txt")

#Plotting Random Forest and result
plot(RF_model)
plot(pred, test[,16], xlab = "Predicted Churn Score", ylab = "Actual Churn Score")

##Evaluate the performance of classification model
ConfMatrix_RF = table(test$Churn, pred)
confusionMatrix(ConfMatrix_RF)

#Naive Bayes
library(e1071)
class_nb = naiveBayes(Churn~., data = train)
predict_nb = predict(class_nb, test[,-16], type = 'class')

#Plotting Naive Bayes result
plot(predict_nb, test[,16], xlab = "Predicted Churn Score", ylab = "Actual Churn Score")

#Evaluating Naive Bayes by confusion matrix
ConfMatrix_nb = table(test$Churn, predict_nb)
confusionMatrix(ConfMatrix_nb)
